### PR TITLE
DAOS-5582 array: fix race condition in dc_array_set_size()

### PR DESCRIPTION
**Pull request by liuxuezhao**
In dc_array_set_size()'s completion callback adjust_array_size_cb(),
    it possibly do two things - tse_task_reinit() and register comp cb by
    tse_task_register_cbs(). In multi-threads executing, after reinit the
    task possibly be scheduled immediately, and register other comp cbs
    ahead by cli_obj IO handing. Then break the different layers' comp
    cbs' order. This patch fix it by register comp cb before task reinit.
    
    Two other changes:
    1) refine the err handling of tse_task_register_comp_cb in obj_task_init
    2) reset task's dtp_stack_top to zero (and print err msg).
    
    Signed-off-by: Xuezhao Liu <xuezhao.liu@intel.com>

### CODE REVIEW

**Review by mchaarawi — DISMISSED** [review]

**Review by mchaarawi — DISMISSED** [review]

**Comment by liuxuezhao** [comment]
just a rebase

**Review by mchaarawi — APPROVED** [review]

**Review by wangdi1 — APPROVED** [review]

File: src/client/array/dc_array.c [hunk 1]
```diff
@@ -2549,6 +2543,12 @@ adjust_array_size_cb(tse_task_t *task, void *data)
 			return rc;
 		}
 
+		rc = tse_task_reinit(task);
+		if (rc) {
+			D_ERROR("FAILED to reinit task\n");
```
Comment (jolivier23) [inline src/client/array/dc_array.c:2548]: DF_RC ?

File: src/common/tse.c [hunk 2]
```diff
@@ -1010,6 +1010,14 @@ tse_task_reinit(tse_task_t *task)
 	dtp->dtp_running = 0;
 	dtp->dtp_completing = 0;
 	dtp->dtp_completed = 0;
+
+	/** reset stack pointer as zero */
+	if (dtp->dtp_stack_top != 0) {
+		D_ERROR("task %p, dtp_stack_top reset from %d to zero.\n",
```
Comment (jolivier23) [inline src/common/tse.c:1016]: what would cause this?

File: src/client/array/dc_array.c (hunk 1 above)
Comment (liuxuezhao) [reply src/client/array/dc_array.c:2548]: ok, can refine it in following patch.

File: src/common/tse.c (hunk 2 above)
Comment (liuxuezhao) [reply src/common/tse.c:1016]: like push to stack but forgot to pop, here just print a err and reset it.

**Pull request merged**